File: py/day02.py
import os

# https://adventofcode.com/2023/day/2

def round_valid(round):
    return round[0] <= 12 and round[1] <= 13 and round[2] <= 14

def rounds_valid(rounds):
    return all([round_valid(round) for round in rounds])

def power(rounds):
    r = max([r for (r,_,_) in rounds])
    g = max([g for (_,g,_) in rounds])
    b = max([b for (_,_,b) in rounds])
    return r*g*b

with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day02-input", "r") as f:
    input = [line.strip() for line in f.readlines()]

games: list[int, ] = []
for line in input:
    (gid, game) = line.split(':')
    rounds = []
    for round in game.split(';'):
        for balls in round.strip().split(','):
            (n,r) = balls.strip().split(' ')
            num = int(n)
            red = num if r == "red" else 0
            green = num if r == "green" else 0
            blue = num if r == "blue" else 0
            rounds.append((red, green, blue))
    games.append((int(gid[5:]), rounds))


# Part 1
p1 = sum([game for (game,rounds) in games if rounds_valid(rounds)])
print("Part 1:", p1)

# Part 2
p2 = sum([power(rounds) for (_,rounds) in games])
print("Part 2:", p2)

File: cs/days/day02.cs
using System.Collections.ObjectModel;
using Spectre.Console.Rendering;

namespace Shunty.AoC;

// https://adventofcode.com/2023/day/2 - Cube Conundrum

public class Day02 : AocDaySolver
{
    public int DayNumber => 2;

    public async Task Solve()
    {
        var input = await AocUtils.GetDayLines(DayNumber);
        var games = new List<Game>();
        foreach (var s in input)
        {
            var gameparts = s.Split(':');
            var gameId = int.Parse(gameparts[0].AsSpan(5));
            var roundparts = gameparts[1].Split(';');
            var rounds = new List<GameRound>();
            foreach (var p in roundparts)
            {
                var rgbparts = p.Split(',');
                foreach (var rgb in rgbparts)
                {
                    var split = rgb.Trim().Split(' ');
                    var num = int.Parse(split[0]);
                    var r = split[1] == "red" ? num : 0;
                    var g = split[1] == "green" ? num : 0;
                    var b = split[1] == "blue" ? num : 0;
                    rounds.Add(new(r,g,b));
                }
            }
            games.Add(new(gameId, rounds));
        }

        int mred = 12, mgreen = 13, mblue = 14;
        var p1 = games
            .Where(g => g.Rounds.All(r => r.Red <= mred && r.Green <= mgreen && r.Blue <= mblue))
            .Sum(g => g.GameId);
        this.ShowDayResult(1, p1);

        var p2 = games.Sum(g =>
            g.Rounds.Max(r => r.Red)
            * g.Rounds.Max(r => r.Green)
            * g.Rounds.Max(r => r.Blue));
        this.ShowDayResult(2, p2);
    }
}

public record GameRound(int Red, int Green, int Blue);
public  record Game(int GameId, ICollection<GameRound> Rounds);
